### DISCUSSION TITLE
Does Tanstack Form uses native Web-API (e.g. checkValidity) for validations?

### DISCUSSION THREAD

**Discussion by nimo23** [TanStack/form#1087] [Ideas] [answered]
Does the TanStackForm `formApi` internally use

`HTMLInputElement.checkValidity(): boolean`
https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/checkValidity

Does the TanStackForm `fieldApi` internally use

`HTMLInputElement.checkValidity(): boolean`
https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/checkValidity

If not, would there be advantages (performance, etc) if `TanStackForm` uses the native APIs here?

Currently, I have an uncontrolled react form:

```
..
function onBlur(e: React.FormEvent) {
    const target = e.target as HTMLInputElement;
    const isValid = target.checkValidity();
    setValidationMessage(isValid ? null : { [target.name]: target.validationMessage });
}

return (
<form noValidate>
<TextField
    name='username'
    defaultValue={user.username}
    type="text"
    required
    onInvalid={onInvalid}
    onBlur={onBlur}
    error={validationMessage?.username != null}
    helperText={validationMessage?.username}
/>
</form>
);
```

which does not make use of TanstackForm-Api. It only uses native Web-Api. It is also possible to customize the native validation message by `field.setCustomValidity(zodMsg)`.

I ask because I'm considering switching to `Tanstack Forms`, but I want to make sure it uses/leverages native web APIs (especially the native client validation API, see https://developer.mozilla.org/en-US/docs/Learn_web_development/Extensions/Forms/Form_validation) for performance and security reasons instead of reinventing the wheel..

**Comment by crutchcorn** [comment, marked as answer]
The native web APIs do not sufficiently cover complex use cases; as such we do not utilize them.

Further, it would intensely tie us to the web, whereas we support React Native and Nativescript as well.

You can glue in TanStack Form into native APIs trivially if you wish, but it's an explicit non-goal for us.
